pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -398,10 +399,17 @@ func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
 		return
 	}
 
-	// First, we need to detach the perf buffer.
-	// We do not check the returned error because if the container was deleted it
-	// was already detached.
-	_ = traceUnique.tracer.Detach(mntNsID)
+	// The kernel can reuse mntNsID, so we want to keep track if a container was
+	// already detached to avoid detaching a non related container.
+	_, alreadyDetached := traceUnique.detachedContainers[mntNsID]
+	if !alreadyDetached {
```

Comment:
We have commit 4dedafc8565dd183b2553f97a046dab52a82d515 from #1284 to force the kernel to keep a mnt ns a little bit longer before releasing it (and therefore being able to reuse it) while we clean up the container.
Could we reuse this mechanism somehow?

It seems that `traceUnique.detachedContainers` is never clean up, so I am not sure this patch works.

cc @mauriciovasquezbernal

Reply:
> Could we reuse this mechanism somehow?

I don't think we need to reuse that. In this case, the trace could be kept a long time after the container is removed, it means we'll keep an open file descriptor for a lot of time. 

I think we can fix this in another way: What about having a detached field that is set to true on `detachContainerFunc` and that is checked in `Delete()`? 

```diff
diff --git pkg/gadget-collection/gadgets/traceloop/gadget.go pkg/gadget-collection/gadgets/traceloop/gadget.go
index 00b1870e..eff02008 100644
--- pkg/gadget-collection/gadgets/traceloop/gadget.go
+++ pkg/gadget-collection/gadgets/traceloop/gadget.go
@@ -33,13 +33,18 @@ import (
        containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
 )
 
+type containerSlim struct {
+       mntnsid  uint64
+       detached bool
+}
+
 type Trace struct {
        client  client.Client
        helpers gadgets.GadgetHelpers
 
        started bool
 
-       containerIDs map[string]uint64
+       containerIDs map[string]*containerSlim
 
        trace *gadgetv1alpha1.Trace
 }
@@ -183,7 +188,7 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
        // The output will contain an array of types.TraceloopInfo.
        // So, to avoid problems, we initialize it to be a JSON array.
        trace.Status.Output = "[]"
-       t.containerIDs = make(map[string]uint64, 0)
+       t.containerIDs = make(map[string]*containerSlim, 0)
 
        genKey := func(container *containercollection.Container) string {
                return container.Namespace + "/" + container.Podname
@@ -203,7 +208,9 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
                        return err
                }
 
-               t.containerIDs[containerID] = mntNsID
+               t.containerIDs[containerID] = &containerSlim{
+                       mntnsid: mntNsID,
+               }
                log.Debugf("tracer attached for %q (%d)", key, mntNsID)
 
                var infos []types.TraceloopInfo
@@ -251,6 +258,10 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
 
                traceUnique.Lock()
                err := traceUnique.tracer.Detach(mntNsID)
+
+               // TODO: check if container.ID exists
+               t.containerIDs[container.ID].detached = true
+
                traceUnique.Unlock()
                if err != nil {
                        log.Errorf("failed to detach tracer: %s", err)
@@ -375,7 +386,7 @@ func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
        }
 
        containerID := trace.Spec.Parameters["containerID"]
-       mntNsID, ok := t.containerIDs[containerID]
+       container, ok := t.containerIDs[containerID]
        if !ok {
                ids := make([]string, len(t.containerIDs))
                i := 0
@@ -401,7 +412,9 @@ func (t *Trace) Delete(trace *gadgetv1alpha1.Trace) {
        // First, we need to detach the perf buffer.
        // We do not check the returned error because if the container was deleted it
        // was already detached.
-       _ = traceUnique.tracer.Detach(mntNsID)
+       if !container.detached {
+               _ = traceUnique.tracer.Detach(container.mntnsid)
+       }
 
        // Then we can remove it.
        err := traceUnique.tracer.Delete(containerID)
```